platform/xen/rumpxendev/evtdev.c

```diff
@@ -0,0 +1,502 @@
+	u_int ring_cons;
+	u_int ring_prod;
+#define EVTDEV_F_OVERFLOW 0x1 	/* ring overflow */
+	u_int flags;
+
+	kmutex_t lock;
+	kcondvar_t cv;
+	struct selinfo selinfo; /* used by poll, see select(9) */
+};
+
+/* Kernel event -> device instance mapping */
+static kmutex_t devevent_lock;
+static struct xenevt_dev_data *devevents[NR_EVENT_CHANNELS];
+
+/*----- helpers -----*/
+#define WBITS (POLLOUT | POLLWRNORM)
+#define RBITS (POLLIN  | POLLRDNORM)
+
+/* call with d->lock held */
+static void queue(struct xenevt_dev_data *d, u_int port)
+{
+	if (d->ring_cons == ((d->ring_prod + 1) & EVTDEV_RING_MASK)) {
+		d->flags |= EVTDEV_RING_MASK;
+		printf("evtdev: ring overflow port %d\n", port);
+	} else {
+		d->ring[d->ring_prod] = port;
+		membar_producer();
+		d->ring_prod = (d->ring_prod + 1) & EVTDEV_RING_MASK;
+	}
+	/* notify */
+	cv_signal(&d->cv);
+	selnotify(&d->selinfo, RBITS, NOTE_SUBMIT);
+}
+
+static void rump_evtdev_callback(u_int port)
+{
+	mutex_enter(&xenevt_ring_lock);
+
+	if (xenevt_ring_cons == ((xenevt_ring_prod + 1) & EVTDEV_RING_MASK)) {
+		minios_printk("xenevt driver ring overflowed!\n");
+	} else {
+		xenevt_ring[xenevt_ring_prod] = port;
+		membar_producer();
+		xenevt_ring_prod = (xenevt_ring_prod + 1) & EVTDEV_RING_MASK;
+	}
+
+	mutex_exit(&xenevt_ring_lock);
+
+	minios_wake_up(&minios_events_waitq);
+}
+
+static void xenevt_thread_func(void *ign)
+{
+	u_int prod = xenevt_ring_prod;
+	u_int cons;
+
+	/* give us a rump kernel context */
+	rumpuser__hyp.hyp_schedule();
+	rumpuser__hyp.hyp_lwproc_newlwp(0);
+	rumpuser__hyp.hyp_unschedule();
+
+	for (;;) {
+		minios_wait_event(minios_events_waitq, xenevt_ring_prod != prod);
+		prod = xenevt_ring_prod;
+		cons = xenevt_ring_cons;
+
+		membar_sync();
+
+		while (cons != prod) {
+			u_int port = xenevt_ring[cons];
+			struct xenevt_dev_data *d;
+
+			KASSERT(port < NR_EVENT_CHANNELS);
+
+			mutex_enter(&devevent_lock);
+
+			d = devevents[port];
+
+			KASSERT(d);
+
+			mutex_enter(&d->lock);
+
+			queue(d, port);
+
+			mutex_exit(&d->lock);
+			mutex_exit(&devevent_lock);
+
+			cons++;
+		}
+
+		membar_sync();
+
+		xenevt_ring_cons = cons;
+	}
+}
+
+/*----- request handling (writes to the device) -----*/
+static int
+xenevt_dev_write(struct file *fp, off_t *offset, struct uio *uio,
+		 kauth_cred_t cred, int flags)
+{
+	struct xenevt_dev_data *d = fp->f_data;
+	uint16_t *chans = NULL;
+	int i, nentries, err;
+
+	DPRINTF(("/dev/xenevt: write...\n"));
+
+	if (uio->uio_resid == 0) {
+		err = 0;
+		goto out;
+	}
+
+	nentries = uio->uio_resid / sizeof(uint16_t);
+	if (nentries > NR_EVENT_CHANNELS) {
+		err = EMSGSIZE;
+		goto out;
+	}
+
+	chans = bmk_memcalloc(nentries, sizeof(uint16_t));
+	if (chans == NULL) {
+		err = ENOMEM;
+		goto out;
+	}
+
+	err = uiomove(chans, uio->uio_resid, uio);
+	if (err) goto out;
+
+	mutex_enter(&devevent_lock);
+	for (i = 0; i < nentries; i++) {
+		if (chans[i] < NR_EVENT_CHANNELS &&
+		    devevents[chans[i]] == d)
+			minios_unmask_evtchn(chans[i]);
```

Comment:
This is the same semantics as NetBSD's xenevt. I would like to stay in line with that.